File: intersect_ll.cpp
//TC: O(n+m), where n and m are lengths of the linked lists
//SC: O(1)

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        
        
        ListNode *tempA = headA;
        ListNode *tempB = headB;
        
        if(tempA == NULL || tempB == NULL)
            return NULL;
        
       
        while(tempA != tempB && tempA != NULL && tempB != NULL){
            
            tempA = tempA->next;
            tempB = tempB->next;   
            
            if(tempA == tempB)
                return tempA;           //handles non intersecting cases as well.
            
            if(tempA == NULL)
                tempA = headB;
            
            if(tempB == NULL)
                tempB = headA;
            
            
            
        }
        
        return tempA;
        
    }
};

File: IntersectLL.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int countA=0;
        int countB=0;
        ListNode cur=headA;
        while(cur!=null){
            countA++;
            cur=cur.next;
        }
        cur=headB;
        while(cur!=null){
            countB++;
            cur=cur.next;
        }
        if(countA>countB){
            ListNode tempA = headA;
            for(int i=0;i<countA-countB;i++){
                tempA=tempA.next;
            }
            ListNode tempB = headB;
            while(tempA!=tempB){
                tempA=tempA.next;
                tempB=tempB.next;
            } return tempA;
        }
        else {
            ListNode tempB = headB;
            for(int i=0;i<countB-countA;i++){
                tempB=tempB.next;
            }
            ListNode tempA = headA;
            while(tempA!=tempB){
                tempA=tempA.next;
                tempB=tempB.next;
            } return tempA;
        }
    }
}